core/include/macros/math.h

```diff
@@ -31,7 +31,7 @@ extern "C" {
                               unsigned int: 1,           \
                               unsigned long: 1,          \
                               unsigned long long: 1,     \
-                              default: (long long)(a) < 0 ? -1 : 1)
+                              default: ((a) > 0LL) ? 1 : -1)
```

Comment:
```suggestion
                              default: ((a) >= 0LL) ? 1 : -1)
```

Or am I missing something? Otherwise this is good to get it

Reply:
```
In file included from /home/benpicco/dev/RIOT/sys/include/embUnit/embUnit.h:47,
                 from /home/benpicco/dev/RIOT/sys/include/embUnit.h:25,
                 from /home/benpicco/dev/RIOT/tests/unittests/tests-core/tests-core-macros.c:11:
/home/benpicco/dev/RIOT/tests/unittests/tests-core/tests-core-macros.c: In function ‘test_math_div_round’:
/home/benpicco/dev/RIOT/core/include/macros/math.h:34:45: error: comparison is always true due to limited range of data type [-Werror=type-limits]
   34 |                               default: ((a) >= 0LL) ? 1 : -1)
      |                                             ^~
/home/benpicco/dev/RIOT/sys/include/embUnit/AssertImpl.h:60:47: note: in definition of macro ‘TEST_ASSERT_EQUAL_INT’
   60 |         long long ____actual__ = (long long) (actual_); \
      |                                               ^~~~~~~
/home/benpicco/dev/RIOT/core/include/macros/math.h:38:33: note: in expansion of macro ‘SIGNOF’
   38 | #define DIV_ROUND(a, b) (((a) + SIGNOF(a) * (b) / 2) / (b))
      |                                 ^~~~~~
/home/benpicco/dev/RIOT/tests/unittests/tests-core/tests-core-macros.c:27:30: note: in expansion of macro ‘DIV_ROUND’
   27 |     TEST_ASSERT_EQUAL_INT(3, DIV_ROUND(a, 7));
      |                              ^~~~~~~~~
cc1: all warnings being treated as errors
```

Reply:
I'd rather have some `#pragma diagnostics` in the macro then inconsistent results with unsigned types

Reply:
But the warning is strictly wrong. It should promote the other side of the comparison to `signed long long`, which can be negative.

Reply:
Note: This is how pragmas work in macros:

https://github.com/RIOT-OS/RIOT/blob/e07747a9f772572347096177831c576dcbc43f9e/core/lib/include/kernel_defines.h#L208-L212

Reply:
does not work:
https://godbolt.org/z/8h84deK9j

Reply:
It seems to be a classical compiler bug: https://godbolt.org/z/bfeEjvx8o

Apparently the warnings are checked for each and every case of the `_Generic` and emitted, and only afterwards the correct case is selected. This yields bogus warnings for cases that are eliminated later on :/ Fuuuuuuck!

Reply:
```suggestion
                              default: ((a) <= 0 ? ((a) == 0 ? 1L : -1L ): 1L))
```
see: https://godbolt.org/z/nE97rKzPe